### PR TITLE
fix: rateLimitDeployment ignoring pod labels and annotation merge

### PR DESCRIPTION
**Pull request by oscarboher**
**What type of PR is this?**
fix
**What this PR does / why we need it**:
`rateLimitDeployment.pod.labels` where not applying to rate limit pods.
prometheus annotations where ignored if `rateLimitDeployment.pod.annotations` was defined

**Which issue(s) this PR fixes**:
Fixes https://github.com/envoyproxy/gateway/issues/4226

### CODE REVIEW

File: internal/infrastructure/kubernetes/ratelimit/resource_provider.go [hunk 1]
```diff
@@ -208,16 +217,16 @@ func (r *ResourceRender) Deployment() (*appsv1.Deployment, error) {
...
 		},
 		Spec: appsv1.DeploymentSpec{
 			Replicas: r.rateLimitDeployment.Replicas,
 			Strategy: *r.rateLimitDeployment.Strategy,
 			Selector: selector,
 			Template: corev1.PodTemplateSpec{
 				ObjectMeta: metav1.ObjectMeta{
-					Labels:      selector.MatchLabels,
-					Annotations: annotations,
+					Labels:      podLabels,
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/ratelimit/resource_provider.go:228]: these are immutable so updating the labels will result in a failed deployment
cc @sanposhiho

Comment (sanposhiho) [reply internal/infrastructure/kubernetes/ratelimit/resource_provider.go:228]: To be accurate, .spec.selector is immutable, while .spec.template.labels isn't. But, we have to make sure the .spec.template.labels match .spec.selector, when updating it.

@oscarboher 
You can refer to the idea -
https://github.com/sanposhiho/gateway/blob/69bf8825422db0cb7b34097b8db5dfca19b8d477/internal/infrastructure/kubernetes/proxy/resource_provider.go#L195-L206
how we can make the selector stable while allowing patching labels.

So, basically, we should build the selector only with the labels that will never get changed. In other words, we shouldn't build the selector with the custom labels that users can change.

Also, you have to pay attention not to break the existing users who already have this deployment on their cluster.

Comment (oscarboher) [reply internal/infrastructure/kubernetes/ratelimit/resource_provider.go:228]: Currently the .spec.selector is a selector with the labels:
```
"app.kubernetes.io/name": InfraName,
"app.kubernetes.io/component":  "ratelimit",
"app.kubernetes.io/managed-by": "envoy-gateway",
```
So i understand this must remain the same to not break things because of the selector immutability.

Isn't it enough to ensure that the podLabels include at least those labels for the selector to work, as my PR does? It does look a bit funny using the copy function but it works.
What am I missing? Or the idea is just to have a method that can be used elsewhere to maintain some selectors?

File: internal/infrastructure/kubernetes/ratelimit/resource_provider.go [hunk 2]
```diff
@@ -186,19 +187,28 @@ func (r *ResourceRender) ServiceAccount() (*corev1.ServiceAccount, error) {
...
 func (r *ResourceRender) Deployment() (*appsv1.Deployment, error) {
 	containers := expectedRateLimitContainers(r.rateLimit, r.rateLimitDeployment, r.Namespace)
-	labels := rateLimitLabels()
-	selector := resource.GetSelector(labels)
+	selector := resource.GetSelector(rateLimitLabels())
+
+	podLabels := rateLimitLabels()
+	if r.rateLimitDeployment.Pod.Labels != nil {
+		maps.Copy(podLabels, r.rateLimitDeployment.Pod.Labels)
+		maps.Copy(podLabels, rateLimitLabels())
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/ratelimit/resource_provider.go:195]: can L195 be removed ? since L192 is already setting it

File: internal/infrastructure/kubernetes/ratelimit/resource_provider.go (hunk 1 above)
Comment (arkodg) [reply internal/infrastructure/kubernetes/ratelimit/resource_provider.go:229]: thanks @sanposhiho and @oscarboher, this change is safe since the `selector` isnt changing

Comment (sanposhiho) [reply internal/infrastructure/kubernetes/ratelimit/resource_provider.go:229]: Oops, sorry, I thought `rateLimitLabels()` included `rateLimitDeployment.pod.labels` too. Yes, then nothing to worry here around the immutability.

File: internal/infrastructure/kubernetes/ratelimit/resource_provider.go (hunk 2 above)
Comment (oscarboher) [reply internal/infrastructure/kubernetes/ratelimit/resource_provider.go:195]: So `maps.Copy(dest, src)` [overwrites keys existing ](https://pkg.go.dev/maps#Copy) in both maps with the `src` value. I want to make sure noone can override one of the selector labels from the `rateLimitLabels`, so I'm doing the last maps.Copy to overwrite with the selector labels in case someone added one of those labels in their values.

Also the destination map must not be nil so I cannot do something like:
```
podLabels := r.rateLimitDeployment.Pod.Labels
maps.Copy(podLabels, rateLimitLabels())
```
as it will fail if no pod labels are provided.
It looks a bit funny, I just didn't want to implement a function that merges two maps, but maybe it's worth it for clarity's sake. Let me know if you think so.

Comment (arkodg) [reply internal/infrastructure/kubernetes/ratelimit/resource_provider.go:195]: ah thats a good way to ensure the labels can't be overridden
nit: this as a code comment would be helpful so the next dev gets the "why"

**Review by arkodg — DISMISSED** [review]
LGTM thanks !

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zirain — APPROVED** [review]
